crates/dojo-defi/src/constant_product_market/components.cairo

```diff
@@ -40,6 +79,189 @@ impl MarketImpl of MarketTrait {
+    // based on the reserves in the market
+    fn quote_amount(self: @Market, quantity: usize) -> u128 {
+        assert(quantity > 0, 'insufficient quantity');
+        assert(self.has_liquidity(), 'insufficient liquidity');
+
+        // Get normalized reserve cash amount and item quantity
+        let (reserve_amount, reserve_quantity) = self.get_reserves();
+
+        // Convert reserve amount and quantity to fixed point
+        let reserve_amount = Fixed::new_unscaled(reserve_amount, false);
+        let reserve_quantity = Fixed::new_unscaled(reserve_quantity, false);
+
+        // Normalize quantity
+        let quantity = quantity.into().try_into().unwrap() * SCALING_FACTOR;
+
+        // Convert quantity to fixed point
+        let quantity = Fixed::new_unscaled(quantity, false);
+
+        // dx = X * dy / Y
+        let amount_optimal = (reserve_amount * quantity) / reserve_quantity;
+        // Convert amount to u128 and unscale
+        amount_optimal.into().try_into().unwrap() / ONE_u128
+    }
+
+    // Inner function to add liquidity to the market, computes the optimal amount and quantity
+    //
+    // Arguments:
+    //
+    // amount: The amount of cash to add to the market
+    // quantity: The quantity of items to add to the market
+    //
+    // Returns:
+    //
+    // (amount, quantity): The amount of cash and quantity of items added to the market
+    fn add_liquidity_inner(self: @Market, amount: u128, quantity: usize) -> (u128, usize) {
+        // If there is no liquidity, then the amount and quantity are the optimal
+        if !self.has_liquidity() {
+            // Ensure that the amount and quantity are greater than zero
+            assert(amount > 0, 'insufficient amount');
+            assert(quantity > 0, 'insufficient quantity');
+            (amount, quantity)
+        } else {
+            // Given the amount, get optimal quantity to add to the market
+            let quantity_optimal = self.quote_quantity(amount);
+            if quantity_optimal <= quantity {
+                // Add the given amount and optimal quantity to the market
+                (amount, quantity_optimal)
+            } else {
+                let amount_optimal = self.quote_amount(quantity);
+                // Ensure that the optimal amount is less than or equal to the given amount
+                assert(amount_optimal <= amount, 'insufficient amount');
+                (amount_optimal, quantity)
+            }
+        }
+    }
+
+    // Add liquidity to the market, mints shares for the given amount of liquidity provided
+    //
+    // Arguments:
+    //
+    // amount: The amount of cash to add to the market
+    // quantity: The quantity of items to add to the market
+    //
+    // Returns:
+    //
+    // (amount, quantity, shares): The amount of cash and quantity of items added to the market and the shares minted
+    fn add_liquidity(self: @Market, amount: u128, quantity: usize) -> (u128, usize, FixedType) {
```

Comment:
This implementation follows the one from [UniswapV2Router02](https://etherscan.io/address/0x7a250d5630b4cf539739df2c5dacb4c659f2488d). Amount of liquidity shares minted/added follows the formula `S = dx * L/X = dy * L/Y` and ensures ratio is intact when liquidity is added